third_party/WebKit/Source/core/loader/FrameLoader.cpp

```diff
@@ -1218,8 +1218,12 @@ String FrameLoader::userAgent() const
     String clientUserAgent = client()->userAgent();
 
     // On Endless, we need to still be able to alter the UserAgent string depending
-    // on the URL we are loading, so we adapt it if needed before applying it.
-    String userAgent = document ? adaptUserAgentForURL(clientUserAgent, document->url()) : clientUserAgent;
+    // on the URL we are loading, so we adapt it if needed before applying it. But we
+    // might not have a committed load, so we first check the URL from the provisional
+    // loader before trying with the regular one, so that we always get a valid URL.
+    const DocumentLoader& activeLoader = m_provisionalDocumentLoader.get() ? *m_provisionalDocumentLoader : *m_documentLoader;
+    const KURL& url = activeLoader.request().url();
+    String userAgent = document ? adaptUserAgentForURL(clientUserAgent, url) : clientUserAgent;
```

Comment:
If this were an upstream review I would say this code belongs in the FrameLoaderClient, but I think this is totally fine for Endless.

I did not check to ensure `m_documentLoader` is guaranteed to be valid if `m_provisionalDocumentLoader` is NULL; I hope that is true?